### PR TITLE
Standalone React and Headless Rails applications

### PR DESCRIPTION
**Pull request by purfectliterature**
This PR splits Coursemology into two apps: a standalone React app and headless Rails app. Both apps are independent of each other and must be bootstrapped on different processes. The default HTML entry point to the app is now the standalone React app.

> [!NOTE]
> At this stage, we cannot afford to make the Rails app [API-only](https://guides.rubyonrails.org/api_app.html). Cookies (and cookie-based sessions), CSRF protection, and more are not available by default in an API-only Rails app. Manual extensions and tests will need to be done to go API-only.

## Development changes
- Use `lvh.me` to access the app. This enables multi-tenancy on local environments. 
Run `host lvh.me` on your Terminal to bind `localhost` and `127.0.0.1` to `lvh.me`. Then, access the React app on `lvh.me:8080` and the Rails app on `lvh.me:5000` (or your set ports). You can now develop on other instances, e.g., `nushigh.lvh.me`.
- CORS is now enforced on the Rails app.
- webpack-dev-server now acts as a reverse proxy to the Rails app. 
It proxies requests with `?format=json` and rewrites the `Origin` header to please CORS. See the `devServer` key on `webpack.dev.js` for more information.
- [Foreman](https://github.com/ddollar/foreman) is now deprecated. 
Bootstrap both React and Rails app on separate processes. This also makes it way easier to use Byebug since [Foreman cannot show standard inputs](https://github.com/ddollar/foreman/issues/703).
- Capybara tests are deprecated. Write new end-to-end tests with Playwright.  
Existing Capybara tests still work, and can be updated or extended if needed as we discover edge cases or bug fixes. But new tests should be written with Playwright.
- `yarn build:test` now sets `BABEL_ENV` to something else than `production`.  
This is done to ensure `data-testid` attributes are not removed during webpack's production build.
- Add a new `.env` in `client` if you haven't already. Then add this entry for reCAPTCHA.  
This is the same [testing key](https://developers.google.com/recaptcha/docs/faq#id-like-to-run-automated-tests-with-recaptcha.-what-should-i-do) as the one in the Rails app. If the site key in Rails app changes, make sure you update this too.
```env
GOOGLE_RECAPTCHA_SITE_KEY=6LeIxAcTAAAAAJcZVRqyHh71UMIEGNQ_MXjiZKhI
```

>[!IMPORTANT]
>We'd probably need to rewrite our Getting Started guide at some point. Maybe sometime after we set up our repo for the Go server.

>[!NOTE]
>Playwright tests currently live in `tests/tests`. The idea is to have ALL end-to-end tests under `tests`. Then we can have `tests/playwright` and `tests/capybara`, for example, and bootstrap them with appropriate `scripts` in `tests/package.json`. I'm not sure if we can strip out Capybara tests from the Rails app, though.
>
> We're going to move some files and folders around once we start writing some Go codes, anyway. So, I envision the root of the repo will have folders like: `client`, `api-v1`, `api-v2`, `tests`, and any other environments/configs/CI files. We can have one script (maybe `package.json`) containing all the configs (environment variables, ports, etc.) at the root of the project.

## CircleCI pipeline changes
- Add `build_client` job to build and cache client.
- Add `test_playwright` job for Playwright tests.
- Node jobs `build_client`, `test_playwright`, `jslint`, and `jstest` now use CircleCI's Node convenience image.
- `jslint` and `jstest` now don't build client since it's unnecessary.
- Yarn cache now also caches `vendor/recorderjs/node_modules`.
- Split `i18n` to `i18n_en` and `i18n_zh`. `i18n_en` is now required for merging to `master`.
- `test` is now `test_rspec`.
- `lint` is now `factorybot_lint`.
- Renamed some commands to be more descriptive.

>[!NOTE]
>After this PR is merged, we need to remove `ci/circleci: test` and `ci/circleci: lint` from [`master`'s branch protection settings](https://github.com/Coursemology/coursemology2/settings/branch_protection_rules/567). Until then, ignore the pending results for `lint` and `test` in below status checks.

## Notable changes
- 🔥 The client is now a fully standalone React app.  
Webpack injects the bundled JavaScript scripts with [`HtmlWebpackPlugin`](https://webpack.js.org/plugins/html-webpack-plugin/). `index.tsx` now immediately binds ReactDOM on script load.
- 🔥 New authentication pages.  
Sign in, Sign out, Forgot password, Reset password, and Resend confirmation email pages.
- 🔥 New Terms of Service and Privacy Policy pages.
These pages are code-split. The contents of these pages are rendered in Markdown. There's a new component, `MarkdownPage`.
- 🔥 Site-wide user options are now split from course user options as an avatar button beside the app logo.
- 🔥 The client app is now split to `UnauthenticatedApp` and `AuthenticatedApp`.  
They are code-split.
- 🔥 `ScribingViewComponent` now lazy loads its `fabric`- and `react-color`-dependent children.  
They are code-split.
- 🔥 We now have a preflight, [`CssBaseline`](https://mui.com/material-ui/react-css-baseline/).  
Some components and pages that were using bare HTML elements, e.g., `h4`, `span`, etc., were converted to their appropriate counterparts with `Typography`.
- 🔥 react-toastify and `NotificationPopup` toasts' styles are now standardised.
- 🔥 `/` when authenticated now routes to a dashboard page: list of courses with and sorted by last accessed times.  
If the user has only one course, it will automatically redirect to that course. If the user has no courses, it will automatically redirect to `/courses`.
- 🔥 Add new `session` store that stores i18n config and authentication state.  
This is intended to be extended further in the future to include authentication details, permissions, actual session IDs, etc., if needed.
- 🔥 The base Axios client now supports CSRF recycling.  
Before making any non-GET requests, it will request a fresh CSRF token and cache it if not already, expired, or invalid. It will continue requesting GET to `/csrf_token` until the non-GET request gets through, for a maximum of 3 times.
- 🔥 The base Axios client now redirects 401 and `CanCan::AccessDenied` 403 to the Sign In and Forbidden pages.
- 🔥 "Attempt" and "Watch" buttons in assessments and videos now directly links to their `/attempt` URLs.  
This way, there's no need for manually GET-ting to `/attempt` and redirect programmatically. This also allows users to Open Link in New Tab on these buttons.
- 🔨 In addition to the React component stack, `ErrorBoundary` now also prints out the raw `Error` stack.
- 🔨 Deprecate `lib/server-context.js`.  
There's no need to infer Rails controller name in React. I18n config and CSRF token are managed by `I18nProvider` and the base Axios client, respectively.
- 🔨 Deprecate `lib/axios.js`.
- 🔨 Footer now appears on short screens on course pages.
- 🔨 `Page` now renders a `Fragment` to reduce extra `div` nesting.
- 🔨 `PopupMenuButton` now supports a `secondary` title/component.
- 🔨 `Accordion` now supports a custom `icon` component.
- 🔨 Capybara: Add `expect_forbidden` matcher.
- 🔨 Capybara: Add new `login_as` and `logout` performers (supersedes Warden's).
- 🔨 Move sass, dotenv-webpack, glob, and mkdirp to `devDependencies`.
- 🐛 Fix clear button focusable in `SearchField`s.
- 🐛 Fix custom `onKeyDown` function is not invocable in `TextField`s.
- 🐛 Fix TypeError when loading `UserEmailSubscriptions` page.
- 🐛 Fix `CoursesIndex` not opening New Course dialog when `?new=true` when loading.

For full changelog, refer to the commit history.

## Deployment changes
- [coursemology-theme](https://github.com/Coursemology/coursemology-theme) can now be safely set as read-only and archived. There's nothing there that is longer relevant to our app.
- [dockerfiles](https://github.com/Coursemology/dockerfiles) may still contain some production-related files and configurations. We'll have to eventually move them here.
- Rewrites must be set for `/static/*`, `/*`, and `?format=json` to rewrite them to the static files, `/index.html`, and the Rails server, respectively.
- CORS configurations and multi-tenancy access for production might still need some manual testing.

### CODE REVIEW

File: client/public/index.html [hunk 1]
```diff
@@ -0,0 +1,13 @@
+<!DOCTYPE html>
+<html lang="en">
+  <head>
```
Comment (ekowidianto) [inline client/public/index.html:3]: Add coursemology icon?

Comment (ekowidianto) [reply client/public/index.html:3]: we have `favicon.html.erb` on the backend side. Perhaps we can move this to fe?

File: app/controllers/attachment_references_controller.rb [hunk 2]
```diff
@@ -10,7 +10,7 @@ def create
   end
 
   def show
-    redirect_to @attachment_reference.url(filename: @attachment_reference.name)
+    render json: { redirectUrl: @attachment_reference.url(filename: @attachment_reference.name) }
```
Comment (ekowidianto) [inline app/controllers/attachment_references_controller.rb:13]: Dont forget to redirect in the frontend side

File: client/app/bundles/users/pages/ConfirmEmailPage.tsx [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+  return (
+    <Widget
+      subtitle={t(translations.emailConfirmedSubtitle, {
+        email,
+        strong: (chunk) => <strong>{chunk}</strong>,
+      })}
+      title={t(translations.emailConfirmed)}
+    >
+      <Widget.Foot>
+        <Typography color="text.secondary" variant="body2">
```
Comment (ekowidianto) [inline client/app/bundles/users/pages/ConfirmEmailPage.tsx:32]: Add a link to login page?

File: client/app/bundles/course/assessment/attemptLoader.ts [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+import { getIdFromUnknown } from 'utilities';
+
+import CourseAPI from 'api/course';
+
+const assessmentAttemptLoader: LoaderFunction = async ({ params }) => {
+  const assessmentId = getIdFromUnknown(params?.assessmentId);
+  if (!assessmentId) return redirect('/');
+
+  const { data } = await CourseAPI.assessment.assessments.attempt(assessmentId);
+  return redirect(data.redirectUrl);
```
Comment (ekowidianto) [inline client/app/bundles/course/assessment/attemptLoader.ts:11]: When the backend cant create a submission and returns an error, it will be caught by the error boundary instead of staying on the page, warn the user to try again?

File: client/app/bundles/course/video/attemptLoader.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+  return `/courses/${courseId}/videos/${videoId}/submissions/${submissionId}/edit`;
+};
+
+const videoAttemptLoader: LoaderFunction = async ({ params }) => {
+  const videoId = getIdFromUnknown(params?.videoId);
+  if (!videoId) return redirect('/');
+
+  const { data } = await CourseAPI.video.submissions.create(videoId);
+
+  const url = data.submissionUrl ?? getSubmissionURL(params, data.submissionId);
```
Comment (ekowidianto) [inline client/app/bundles/course/video/attemptLoader.ts:23]: Perhaps, the backend only needs to return the submissionId and no need for submissionUrl anymore? This way we dont need to do extra check if submissionUrl exists in the frontend

File: client/app/lib/constants/sharedConstants.ts [hunk 6]
```diff
@@ -61,3 +61,8 @@ export default {
   STAFF_ROLES,
   AVAILABLE_LOCALES,
 };
+
+export const SUPPORT_EMAIL = 'coursemology@gmail.com' as const;
```
Comment (ekowidianto) [inline client/app/lib/constants/sharedConstants.ts:65]: I am thinking to put all these configurable (imagine when nusri wants to change these next time) as env constants. For locale and time zone, if the values do not exists in the env file, then can fall back to `en` and `Asia/Singapore`.

File: client/app/bundles/users/pages/SignUpLandingPage.tsx [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+        <Link to="/users/sign_in">{t(translations.signIn)}</Link>
+      </Widget.Foot>
+
+      <Widget.Foot className="flex space-x-3">
+        <Typography color="text.secondary" variant="body2">
+          {t(translations.didntReceiveConfirmationEmail)}
+        </Typography>
+
+        <Link to="/users/confirmation/new">
```
Comment (ekowidianto) [inline client/app/bundles/users/pages/SignUpLandingPage.tsx:38]: For this one, could it be triggered automatically, stay on the page, and disable resending the confirmation email for another 10 seconds to avoid spamming?

File: client/app/bundles/common/DashboardPage.tsx [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+    defaultMessage: "Oops, no courses matched ''{keyword}''.",
+  },
+});
+
+const DashboardPage = (): JSX.Element => {
+  const { courses } = useAppContext();
+
+  const { t } = useTranslation();
+
+  const [filterKeyword, setFilterKeyword] = useState('');
```
Comment (ekowidianto) [inline client/app/bundles/common/DashboardPage.tsx:38]: Reuse useItems for filtering and sorting?

File: client/app/bundles/common/DashboardPage.tsx [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+      <SearchField
+        autoFocus
+        onChangeKeyword={setFilterKeyword}
+        placeholder={t(translations.searchCourses)}
+      />
+
+      {Boolean(courses?.length) && (
+        <section className="flex flex-col space-y-5">
+          {filteredCourses?.map((course) => (
+            <Link
```
Comment (ekowidianto) [inline client/app/bundles/common/DashboardPage.tsx:69]: Separate this out?

File: client/app/bundles/common/ErrorPage.tsx [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+  illustrationSrc: string;
+  illustrationAlt: string;
+  title: ReactNode;
+  subtitle: ReactNode;
+  attributions?: Attributions;
+  tip?: ReactNode | false;
+  children?: ReactNode;
+}
+
+const ErrorPage = (props: ErrorPageProps): JSX.Element => {
```
Comment (ekowidianto) [inline client/app/bundles/common/ErrorPage.tsx:83]: Love these pages!

**Review by ekowidianto — CHANGES_REQUESTED** [review]
@purfectliterature Really great work for pushing this thru and thanks! Few comments and lets try to deploy soon. I will look through the test-related code separately and the comments below and above can be addressed first. 
- [ ] When I am not logged in, closing a global announcement should not redirect me to sign in page
- [ ] Can I view course show page when not signed in?
- [ ] Maybe we dont need to see the "see all public courses" when i am already in the course index page?
image
- [ ] ~~Could you also handle 404, 413, 422, 500 & 504 error codes? If possible abstract and generalise the code (isUnauthenticatedResp, isUnauthorizedRes, etc) so that we can add more error code handling easier in the future~~  
→ You'd probably be imagining a map that maps the HTTP error codes to a function. This is not possible because our Axios client redirects not only based on the HTTP error code, but also some other criteria (usually the error message). Overengineering the interceptors will just lead to excessive function calls for every request. The idea is not to crowd the Axios client with too many handlers. It will lead to unexpected results in the components, i.e., try-catch no longer working because the interceptors' redirects.
- [ ] Also, replace render file to render json in the backend and move/port the items in the backend `public` folder.
- [ ] File download not working
- [ ] Could you help fixing the failed test cases? These test cases failed as I removed the html response from the controllers
- [ ] ~~Add playwright testing guide~~  
→ There's nothing particularly noteworthy about writing Playwright tests. Just add the respective tests in the `tests/tests` folder. Then again, we need to see how the Playwright tests scale as we go from now. If we're going to write anything, I'd rather write it together when we update our Getting Started guide, as I mentioned in my PR description above.

File: client/app/bundles/users/components/Widget.tsx [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+  className?: string;
+}
+
+interface WidgetProps extends ContainerProps {
+  title: string;
+  subtitle?: string;
+}
+
+const Widget = (props: WidgetProps): JSX.Element => (
+  <form
```
[the remark this answers is not in the document]
Comment (purfectliterature) [reply client/app/bundles/users/components/Widget.tsx:15]: Because there's no need to. It's just a couple of fields and `useState` is enough for the job. Using `Form` brings so much burden of react-hook-form, re-renders for every value change, etc.

File: client/app/bundles/users/pages/SignUpLandingPage.tsx (hunk 7 above)
Comment (purfectliterature) [reply client/app/bundles/users/pages/SignUpLandingPage.tsx:38]: What's the point? Someone could just refresh the page and make another request to resend a confirmation email.

I specifically didn't want to do anything about debouncing this unless the solution comes from the server.

Comment (ekowidianto) [reply client/app/bundles/users/pages/SignUpLandingPage.tsx:38]: Yeah we should do server-side debouncing. Lets do this when working on the authentication porting next

File: client/app/bundles/common/DashboardPage.tsx (hunk 8 above)
Comment (purfectliterature) [reply client/app/bundles/common/DashboardPage.tsx:38]: I don't think it's useful in this case. There's no pagination in this page.

File: client/app/lib/constants/sharedConstants.ts (hunk 6 above)
Comment (purfectliterature) [reply client/app/lib/constants/sharedConstants.ts:65]: I don't think environment variables are the right place to put these constants. Environment variables are ephemeral. I'd rather put these in `package.json`. That's the best place to put it since these values are not ephemeral, but per-build-based.

Then again I don't think NUSRI will want to use our client app.

File: client/app/bundles/common/DashboardPage.tsx (hunk 8 above)
Comment (ekowidianto) [reply client/app/bundles/common/DashboardPage.tsx:38]: It is not necessary to use the pagination feature, you can just pass the whole length of items to the hook.

Comment (ekowidianto) [reply client/app/bundles/common/DashboardPage.tsx:38]: You will start to have repeated code in the codebase (ie DashboardPage, DayView, etc)

File: client/app/lib/constants/sharedConstants.ts (hunk 6 above)
Comment (ekowidianto) [reply client/app/lib/constants/sharedConstants.ts:65]: What do you mean by environment variables being ephemeral? So lets say if the support email is changed, or someone else uses our client, the app needs to be rebuilt and redeployed every time?

File: app/controllers/attachment_references_controller.rb (hunk 2 above)
Comment (purfectliterature) [reply app/controllers/attachment_references_controller.rb:13]: There's actually nothing wrong about having `redirect_to` here. Attachments will ultimately go to `/uploads/*` anyway. It's the reverse proxy's job to proxy `/attachments/*` and `/uploads/*` to the server for downloads to work.

**Comment by ekowidianto** [comment]
`You'd probably be imagining a map that maps the HTTP error codes to a function. This is not possible because our Axios client redirects not only based on the HTTP error code, but also some other criteria (usually the error message). `
Wouldnt it be possible also then to pass the error message to the function returned by the map?

`Overengineering the interceptors will just lead to excessive function calls for every request.`
Wouldnt the error functions only be called when there's an error?

`The idea is not to crowd the Axios client with too many handlers. It will lead to unexpected results in the components, i.e., try-catch no longer working because the interceptors' redirects.`
The least we should be handling the error codes mentioned above. (ie when a course component such as achievement, etc is not found, our backend returns 404 to the frontend and it is not handled right now). You are right that it would lead to unexpected results when devs just keep adding new error codes but they should proceed with caution. In fact almost all errors that supposed to be caught by the try-catch block is of 400 code which we dont intercept.

**Comment by ekowidianto** [comment]
`There's nothing particularly noteworthy about writing Playwright tests. Just add the respective tests in the tests/tests folder. Then again, we need to see how the Playwright tests scale as we go from now. If we're going to write anything, I'd rather write it together when we update our Getting Started guide, as I mentioned in my PR description above.`

Could you also update the readme and the getting started guide based on what we have now? We wont be getting to Go that soon (maybe in 1-2 months time) and we just need to add go-related guide in due time.

File: client/app/lib/constants/sharedConstants.ts (hunk 6 above)
Comment (purfectliterature) [reply client/app/lib/constants/sharedConstants.ts:65]: Yes. They should change those values and commit it to their version control.

**Comment by purfectliterature** [comment]
@ekowidianto

>Can I view course show page when not signed in?

No. We have never supported that. A user need not sign in to view all courses, but must be signed in to view any courses. You can check out to c9dec948c and verify this behaviour.

>Could you also handle 404, 413, 422, 500 & 504 error codes?

As discussed, we will not be intercepting 404 responses.

I don't think we have explicit 422's. Generally, 422 should be thrown for invalid CSRF tokens, but our app throws 403 instead and it has since been handled.

I don't think we can add redirects for 5XX responses because then the nice feature of allowing the client to load optimistically when the server is down, introduced in eb54a82, will not work. If the user already have some page loaded, then initiate an API request, and the server responds 5XX, then it will either (a) pop up a toast, if handled, or (b) default to `ErrorBoundary`. I think this is reasonable. Case (a) allows the browser session to be uninterrupted. Redirects for case (b) is nothing more than just cosmetic.

I will add a redirect for 404 due to `ComponentNotFound` error.

**Comment by purfectliterature** [comment]
The Playwright tests keep failing on CircleCI for reasons God knows why, regardless of being run headed or headless. They all pass on my computer though.

I think we can skip figuring out why CircleCI keeps failing our Playwright tests for now.

**Review by ekowidianto — APPROVED** [review]
LGTM! Lets try deploying it on staging! @purfectliterature
